### ISSUE TITLE
Can't assign mouse buttons to editor shortcuts

### ISSUE THREAD

**Issue by EricEzaM** [godotengine/godot#38326]
**Godot version:**
All versions including v4.0.dev.custom_build.1d45a269f
**OS/device including version:**
W10 1903
**Issue description:**
Can't assign mouse button inputs to 
**Steps to reproduce:**
Editor Settings > Shortcuts > Try to set any mouse button, can't
**Minimal reproduction project:**
None

Some related issues:
#6366
#26999

The editor settings only allows key inputs, as shown below, where the event is cast to InputEventKey.
https://github.com/godotengine/godot/blob/f6e29addd433130100061935b9a2b8f6de38254b/editor/settings_config_dialog.cpp#L318-L330

I think being able to assign BUTTON_XBUTTON1 and BUTTON_XBUTTON2 would be helpful, especially for back/forward editor navigation. In order to not let people mess up the general usage of the editor, I think it may be a case of only allowing the above 2 mouse button input types to be assigned... i.e. checking that the input event is one of the above, and not allowing rebinding of MB1, MB2, scroll, etc (unless they are modified...? e.g. with shift, alt, ctrl).

Making this change would be relatively simple (I think - haven't looked into it too deep), but it would require changing a decent amount of existing code within the file above, since `last_wait_for_key` would need it's type altered.

**Comment by Calinou** [comment]
See also #36910, where I had to hardcode the keys due to this limitation (in addition to not being able to assign multiple shortcuts to the same action).

**Comment by KoBeWi** [comment]
Duplicate of #6648 basically, but it's more detailed.

**Comment by EricEzaM** [comment]
Yeah so I think the issue with the solutions mentioned so far is that they limit flexibility. If the button is hard-coded then people who use those for Push To Talk wouldn't be happy, as some have already expressed in those issues.

This would allow for more opions, as those people could even rebind the key to ctrl/alt/shit + Mouse Button X in order to have both functions available to them.

I will start to implement this change and keep this issue up to date if I run into any hurdles... if not, expect a PR soon :D

**Comment by EricEzaM** [comment]
@Calinou @KoBeWi So I pretty much have this done (shortcuts working - tested after assigning):

![godot windows tools 64_XOADxDieJF](https://user-images.githubusercontent.com/41730826/80714969-09adc000-8b39-11ea-8ca1-5766635b2f1b.png)

Only one issue, as below. The shortcuts use `as_text()` on the InputEvent to get the display information. Fine for keyboard... not so fine for mouse, as you can see.
![godot windows tools 64_hpMSfAhq7D](https://user-images.githubusercontent.com/41730826/80715120-41b50300-8b39-11ea-857c-5fb826fdc466.png)

I have thought about solutions and have considered:
1. Adding another virtual method to InputEvent to get the text (not preferred)
2. Changing `as_text()` to `as_text(bool p_readable)` or something similar. Where `p_readable=false` is what it gives now, and `p_readable = true` would give the same thing given in the first image above. The exact wording of the parameter could be anything of course.

As far as I can tell `as_text()` is only used in shortcut.cpp, in `get_as_text()` so it would not have wide reaching changes.

Keen to hear thoughts.

**Comment by Calinou** [comment]
@EricEzaM I think it would be better to make `as_text()` return an human-readable string for all input event types. If we still need a debugging-friendly conversion, it could be done as an implicit `_to_string()` conversion (like array/dictionary types).

**Comment by EricEzaM** [comment]
Bit of an update...

Using keys only for shortcuts is quite deeply ingrained in the engine code... Most checks for shortcuts happen on `_unhandled_key_input(...)` which clearly is not helpful when it comes to implementing non-key shortcuts. 

So basically everywhere where shortcuts are read (so far only seems to be <5 classes) key input needs to be changed to check for mouse button input also. PR to come soon I hope.

**Comment by EricEzaM** [comment]
FYI reduz is against this change currently due to fundamental differences bwteen how KB and Mouse input is handled.

Or at least he was several months ago when this was discussed on IRC.

**Comment by me2beats** [comment]
I think this is a useful feature.
Without going into implementation details, I can assume that for the end user, keyboard keys and mouse buttons are all buttons.
And assigning actions to mouse buttons is a fairly common practice.

If it is difficult to implement within the current editor's shortcut system architecture, then work-rounds could be provided, for example, ability to create a custom/additional shortcut system.

btw, what about something similar to InputMap?
That is, all shortcuts assignments could be stored in EditorSetting, etc.

**Comment by MagdielM** [comment]
I'd find this particularly useful to rebind some rather irritating bindings that I can't change as of now. At least for my mouse, tilting the mouse wheel is bound to panning the graph editor, which happens often on accident when I try to click the mouse wheel in to pan by dragging. Often, this ends up happening:
https://user-images.githubusercontent.com/56076033/108091880-82dbbc80-7052-11eb-8b1b-f8db70a4565b.mp4

**Comment by Calinou** [comment]
@MagdielM If this happens to you, it might be worth using dedicated software to disable the lateral mouse wheel buttons.

**Comment by Dok11** [comment]
I would like to set shortcut for next/prev editor tabs via Ctrl+Shift+MouseWheelUp/Down. Does this issue relative for this?

**Comment by Calinou** [comment]
> I would like to set shortcut for next/prev editor tabs via Ctrl+Shift+MouseWheelUp/Down. Does this issue relative for this?

Yes, this issue is about any editor shortcut that can be configured. These shortcuts can be configured:

![Uploading image.png…]()
